### PR TITLE
Sw 131/data recording UI

### PR DESCRIPTION
**Pull request by jrvollmer**
**Changes:**
- Integrated data recording into the front end.
  - Popover trigger for data recording controls can be dragged anywhere within the dashboard.
- Removed code for replaying recorded data.
- Fixed bug with recorded data being bad due to bad incoming packets.
  - Checking size of incoming packets to ensure all necessary data is included. If not, the bad packet is ignored.
- Implemented recording of solar_car_connection (status of comms with solar car).
- Updated README to reflect Python dependencies and location of recorded data.
- Ignore recorded data and list of sessions.

**Known bugs/limitations:**
- Data recording controls open when mouse is released after dragging the popover trigger.

**Notes for reviewer:**
- Please create a session, record data, and then convert that data to Excel using the dashboard. Then, run `git status` and check that there are no changes.
  - Notably, `/Backend/src/routes/*.csv`, `/Backend/recordedData/sessions/*.bin`, and `/Backend/recordedData/sessionsList.bin` should not appear as modified or new to git. However, there should be new .bin and .csv files in these locations on your local copy of the project.

### CODE REVIEW

File: Backend/src/routes/api.js [hunk 1]
```diff
@@ -100,200 +103,73 @@ ROUTER.post("/record-data", (req, res) => {
...
-  console.log(currentSession);
+  if (currentSession === "") {
+    res.send({ response: "NoFileSelected" }).status(200);
+    return
+  }
 
-  // TODO const python = spawn('python', ['./src/routes/exp_new_script1.py','./src/routes/demo3.bin','./Data/sc1-data-format/format.json','./src/routes/test.csv']);
-  const python = spawn('python', ['./src/routes/exp_new_script1.py',
+  // Spawn new child process to call the python script
+  const python = spawn('python', ['./src/routes/process_recorded_data.py',
```
Comment (EUdds) [inline Backend/src/routes/api.js]: we should check for python version here or enforce a dockerized enviornment here. python doesn't mean python for everyone else

File: Backend/src/routes/api.js [hunk 2]
```diff
@@ -100,200 +103,73 @@ ROUTER.post("/record-data", (req, res) => {
...
-  }
-  getrecordedData()
 
-  res.send({ response: RecodedData }).status(200);
-})
+function recordData(data) {
+  fs.appendFile("recordedData/sessions/" + currentSession + ".bin", Buffer.concat([data, Buffer.alloc(1, true)]),
+        (err) => {
+                  if(err) {
+                    console.warn("ERROR: Error while appending to file");
```
Comment (EUdds) [inline Backend/src/routes/api.js]:
```suggestion
                    console.error("ERROR: Error while appending to file");
```

File: Backend/src/routes/api.js [hunk 3]
```diff
@@ -344,6 +211,23 @@ function openSocket() {
...
+                  if(!err) {
+                    // Successfully wrote byte to offset
+                  } else {
+                    console.warn("ERROR: Error writing to file when lost connection");
+                  }
+                }
+            );
+          }
+        });
+      }
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Lets use a if(err) exit kind of structure here. It's hard to follow. Instead of having the error condition at the bottom you should have it at the top

File: Backend/src/routes/process_recorded_data.py [hunk 4]
```diff
@@ -1,7 +1,6 @@
 import sys
```
Comment (EUdds) [inline Backend/src/routes/process_recorded_data.py]: apply black formatting to file. https://black.readthedocs.io/en/stable/

File: DataGenerator/src/server.js [hunk 5]
```diff
@@ -103,7 +103,7 @@ const SERVER = NET.createServer((socket) => {
     }
 
     socket.write(buf1);
-  }, 1000); // TODO 34
+  }, 67); // TODO 67
```
Comment (EUdds) [inline DataGenerator/src/server.js]: What does TODO 67 mean?

**Review by EUdds — CHANGES_REQUESTED** [review]
First pass done.

File: Backend/src/routes/api.js (hunk 1 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Yeah, the plan is to match the docker environment. Then, I can add some instructions to the README to alias `python` (or `python3` or `py` or whatever it happens to be) to refer to an appropriate version of Python 3.X.X.

File: Backend/src/routes/api.js (hunk 2 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: Backend/src/routes/api.js (hunk 3 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: DataGenerator/src/server.js (hunk 5 above)
Comment (jrvollmer) [reply DataGenerator/src/server.js]: Nothing. The TODO was the original frequency for generating random data. It's not important though, so I removed the TODO

File: Backend/src/routes/api.js [hunk 6]
```diff
@@ -1,35 +1,21 @@
...
-let lineNumber = 1;
-
-// Getting all the created sessions from sessionsList.bin
-while (line = broadbandLines.next()) {
-  // console.log(`Line ${lineNumber} has: ${line.toString('ascii')}`);
-  sessionsList.push(line.toString('utf8').replace('\r', ""))
-  lineNumber++;
+let bytesPerPacket = 0;
+for (const property in DATA_FORMAT) {
+  bytesPerPacket += DATA_FORMAT[property][0];
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Change 0 to be a const labelling why it's the 0th index

File: Backend/src/routes/api.js [hunk 7]
```diff
@@ -39,16 +25,34 @@ ROUTER.get("/api", (req, res) => {
...
+let sessionsList = [];
+
+const broadbandLines = new nReadlines('./recordedData/sessionsList.bin');
+let line;
+let lineNumber = 1;
+
+// Getting all the created sessions from sessionsList.bin
+while (line = broadbandLines.next()) {
+  // console.log(`Line ${lineNumber} has: ${line.toString('ascii')}`);
+  sessionsList.push(line.toString('utf8').replace('\r', ""))
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Let's make this a clean function

File: Backend/src/routes/api.js [hunk 8]
```diff
@@ -39,16 +25,34 @@ ROUTER.get("/api", (req, res) => {
...
+  lineNumber++;
+}
+console.log(sessionsList)
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Label console log

File: Backend/src/routes/api.js [hunk 9]
```diff
@@ -344,6 +211,23 @@ function openSocket() {
     if (solarCarData.solar_car_connection.length > 0) {
       solarCarData.solar_car_connection[0] = false;
       frontendData.solar_car_connection[0] = false;
+      // If recording, replace the latest solar_car_connection value in file with false
+      if(doRecord) {
+        fs.open("recordedData/sessions/" + currentSession + ".bin", "r+", (err, fd) => {
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Make string paths a const

File: Backend/src/routes/api.js [hunk 10]
```diff
@@ -344,6 +211,23 @@ function openSocket() {
...
+          if(!err) {
+            fs.write(
+                fd, Buffer.alloc(1, false), 0, 1, fs.fstatSync(fd).size - 1,
+                (err, bw, buf) => {
+                  if(!err) {
+                    // Successfully wrote byte to offset
```
Comment (EUdds) [inline Backend/src/routes/api.js]: Do if err not if !err else

**Review by EUdds — CHANGES_REQUESTED** [review]

File: Backend/src/routes/process_recorded_data.py (hunk 4 above)
Comment (jrvollmer) [reply Backend/src/routes/process_recorded_data.py]: Done

File: Backend/src/routes/api.js (hunk 10 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed in most recent commit. Sorry for the premature comments on the previous review.

File: Backend/src/routes/api.js (hunk 7 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: Backend/src/routes/api.js (hunk 8 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: Backend/src/routes/api.js (hunk 6 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: Backend/src/routes/api.js (hunk 9 above)
Comment (jrvollmer) [reply Backend/src/routes/api.js]: Addressed

File: Backend/src/routes/process_recorded_data.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+import time
+import json
+import numpy as np
+import os
+import xlsxwriter
+
+
+print("#Hello from python#")
+print("First param: " + sys.argv[1])
+print("Second param: " + sys.argv[2])
```
Comment (EUdds) [inline Backend/src/routes/process_recorded_data.py]: Still necessary?

**Review by EUdds — APPROVED** [review]
one nit

File: Backend/src/routes/process_recorded_data.py (hunk 11 above)
Comment (jrvollmer) [reply Backend/src/routes/process_recorded_data.py]: Nope. Removed
